File: array/39.py
#backtracking
class Solution:
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        ret = []
        candidates.sort()
        self.backtrack(ret, [], candidates, target, 0)
        return ret
    
    def backtrack(self, ret, temp, candidates, value, start):
        if value < 0:
            return
        elif value == 0:
            ret.append(temp)
            return
        else:
            for i in range(start, len(candidates)):
                self.backtrack(ret, temp + [candidates[i]], candidates, value - candidates[i], i)
                
    
        

File: backtracking/39.cc
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        std::vector<std::vector<int>> ret;
        std::vector<int> cur;
        dfs(target, 0, ret, cur, candidates);
        return ret;
    }
private:
    void dfs(int target, int sum, std::vector<std::vector<int>>& ret, std::vector<int>& cur, std::vector<int>& cands) {
        if (sum > target)
            return;
        if (sum == target) {
            ret.emplace_back(cur);
            return;
        }
        for (auto i : cands) {
            cur.emplace_back(i);
            dfs(target, sum+i, ret, cur, cands);
            cur.pop_back();
        }
        return;
    }
};
